libfreefare/freefare.3
.\" Copyright (C) 2010 Romain Tartiere
.\"
.\" This program is free software: you can redistribute it and/or modify it
.\" under the terms of the GNU Lesser General Public License as published by the
.\" Free Software Foundation, either version 3 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but WITHOUT
.\" ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
.\" FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
.\" more details.
.\"
.\" You should have received a copy of the GNU Lesser General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>
.\"
.Dd March 30, 2010
.Dt FREEFARE 3
.Os
.\"  _   _
.\" | \ | | __ _ _ __ ___   ___
.\" |  \| |/ _` | '_ ` _ \ / _ \
.\" | |\  | (_| | | | | | |  __/
.\" |_| \_|\__,_|_| |_| |_|\___|
.\"
.Sh NAME
.Nm freefare_get_tags ,
.Nm freefare_get_tag_type ,
.Nm freefare_get_tag_friendly_name ,
.Nm freefare_get_tag_uid ,
.Nm freefare_set_tag_timeout ,
.Nm freefare_free_tag ,
.Nm freefare_free_tags ,
.Nm freefare_version
.Nd Generic target manipulation functions
.\"  _     _ _
.\" | |   (_) |__  _ __ __ _ _ __ _   _
.\" | |   | | '_ \| '__/ _` | '__| | | |
.\" | |___| | |_) | | | (_| | |  | |_| |
.\" |_____|_|_.__/|_|  \__,_|_|   \__, |
.\"                               |___/
.Sh LIBRARY
Mifare card manipulation library (libfreefare, \-lfreefare)
.\"  ____                              _
.\" / ___| _   _ _ __   ___  _ __  ___(_)___
.\" \___ \| | | | '_ \ / _ \| '_ \/ __| / __|
.\"  ___) | |_| | | | | (_) | |_) \__ \ \__ \
.\" |____/ \__, |_| |_|\___/| .__/|___/_|___/
.\"        |___/            |_|
.Sh SYNOPSIS
.In freefare.h
.Ft "FreefareTag *"
.Fn freefare_get_tags "nfc_device_t *device"
.Bd -literal
enum freefare_tag_type {
    FELICA,
    MIFARE_MINI
    MIFARE_CLASSIC_1K,
    MIFARE_CLASSIC_4K,
    MIFARE_DESFIRE,
    MIFARE_ULTRALIGHT,
    MIFARE_ULTRALIGHT_C,
};
.Ed
.Ft "enum freefare_tag_type"
.Fn freefare_get_tag_type "FreefareTag tag"
.Ft "const char *"
.Fn freefare_get_tag_friendly_name "FreefareTag tag"
.Ft "char *"
.Fn freefare_get_tag_uid "FreefareTag tag"
.Ft "void"
.Fn freefare_set_tag_timeout "FreefareTag tag" "int timeout"
.Ft "void"
.Fn freefare_free_tag "FreefareTag tags"
.Ft "void"
.Fn freefare_free_tags "FreefareTag *tags"
.Ft "const char *"
.Fn freefare_version
.\"  ____                      _       _   _
.\" |  _ \  ___  ___  ___ _ __(_)_ __ | |_(_) ___  _ __
.\" | | | |/ _ \/ __|/ __| '__| | '_ \| __| |/ _ \| '_ \
.\" | |_| |  __/\__ \ (__| |  | | |_) | |_| | (_) | | | |
.\" |____/ \___||___/\___|_|  |_| .__/ \__|_|\___/|_| |_|
.\"                             |_|
.Sh DESCRIPTION
The
.Fn freefare_*
family of functions allow agnostic access to the
.Vt FreefareTag
present on a given NFC device.
.Pp
.Fn freefare_get_tags
returns a NULL-terminated list of
.Vt FreefareTag
present on
.Vt device .
This list has to be freed after usage by either:
.Bl -hyphen
.It
calling the
.Fn freefare_free_tags
function.  All tags in the list are automatically freed;
.It
calling the
.Fn freefare_free_tag
function to free each tag in the list individually, and freeing the list itself
using the
.Fn free
function.
.El
.Pp
Because of the nature of the target detection process, any previously detected
target will be in an inconsistent state after a call to
.Fn freefare_get_tags .
It is the programmer's responsibility to free these targets before calling the
.Fn freefare_get_tags
function.
.Pp
Information about a given
.Vt FreefareTag
can be gathered using the
.Fn freefare_get_tag_type ,
.Fn freefare_get_tag_uid
and
.Fn freefare_get_tag_friendly_name
functions.
.Pp
The
.Fn freefare_set_tag_timeout
function is used to set the maximum duration for a single NFC operation to
.Fa timeout
mili-seconds. Setting
.Fa timeout
to 0 disables the timeout feature. By default, a timeout of 2000 is configured.
.Fn freefare_version
function returns the version of the library.
.\"  ____      _                                 _
.\" |  _ \ ___| |_ _   _ _ __ _ __   __   ____ _| |_   _  ___  ___
.\" | |_) / _ \ __| | | | '__| '_ \  \ \ / / _` | | | | |/ _ \/ __|
.\" |  _ <  __/ |_| |_| | |  | | | |  \ V / (_| | | |_| |  __/\__ \
.\" |_| \_\___|\__|\__,_|_|  |_| |_|   \_/ \__,_|_|\__,_|\___||___/
.\"
.Sh RETURN VALUES
Unless stated otherwise, all functions return a value greater than or equal to
.Va 0
on success or
.Va -1
on failure.
.\"  ____                    _
.\" / ___|  ___  ___    __ _| |___  ___
.\" \___ \ / _ \/ _ \  / _` | / __|/ _ \
.\"  ___) |  __/  __/ | (_| | \__ \ (_) |
.\" |____/ \___|\___|  \__,_|_|___/\___/
.\"
.Sh SEE ALSO
.Xr free 3 ,
.Xr mifare_classic 3 ,
.Xr mifare_ultralight 3
.\"     _         _   _
.\"    / \  _   _| |_| |__   ___  _ __ ___
.\"   / _ \| | | | __| '_ \ / _ \| '__/ __|
.\"  / ___ \ |_| | |_| | | | (_) | |  \__ \
.\" /_/   \_\__,_|\__|_| |_|\___/|_|  |___/
.\"
.Sh AUTHORS
.An Romain Tartiere Aq romain@blogreen.org
.An Romuald Conty Aq romuald@libnfc.org


NEWS
Changes between 0.3.4 and 0.3.5  [20 jan 2013]

  *) Depends on libnfc 1.7.0

Changes between 0.3.3 and 0.3.4  [1 mar 2012]

  *) Fix Mifare Ultralight C card detection routine

Changes between 0.3.2 and 0.3.3  [1 mar 2012]

  *) Depends on libnfc 1.6.0

Changes between 0.3.1 and 0.3.2  [17 oct 2011]

  *) New API functions to ease up creation of applications with strong
     cryptography on Mifare DESFire EV1:
     mifare_desfire_create_application_3k3des(),
     mifare_desfire_create_application_aes().
  *) New API functions for ISO 7616 support om Mifare DESFire EV1:
     mifare_desfire_create_application_iso(),
     mifare_desfire_create_application_3k3des_iso(),
     mifare_desfire_create_application_aes_iso(),
     mifare_desfire_get_df_names(),
     mifare_desfire_get_iso_file_ids(),
     mifare_desfire_create_std_data_file_iso(),
     mifare_desfire_create_backup_data_file_iso(),
     mifare_desfire_create_linear_record_file_iso(),
     mifare_desfire_create_cyclic_record_file_iso().
  *) New public API function freefare_tag_new();
  *) Microsoft Windows support.

Changes between 0.3.0 and 0.3.1  [23 feb 2011]

  *) Fix mifare_classic_transfer() for devices returning a 1 byte response on
     success.
  *) New API function mifare_desfire_last_pcd_error().

Changes between 0.2.3 and 0.3.0  [23 dec 2010]

  *) Add support for ISO and AES authentication through
     mifare_desfire_authenticate_iso() and mifare_desfire_authenticate_aes().
  *) Add support for 3K3DES and AES cryptographic operations and add new
     functions mifare_desfire_3k3des_key_new(),
     mifare_desfire_3k3des_key_new_with_version(), mifare_desfire_aes_key_new()
     and mifare_desfire_aes_key_new_with_version() for generating keys.
  *) New functions mifare_desfire_free_mem(),
     mifare_desfire_set_configuration(), mifare_desfire_set_default_key(),
     mifare_desfire_set_ats(), mifare_desfire_get_card_uid(), for Mifare
     DESFire EV1 targets manipulation.
  *) Deprecate authentication information when deleting the currently selected
     application

Changes between 0.2.2 and 0.2.3  [23 dec 2010]

  *) Fix a buffer overflow in mifare_desfire_get_file_settings().

Changes between 0.2.1 and 0.2.2  [23 nov 2010]

  *) Fix build when cutter is not available or is disabled.

Changes between 0.2.0 and 0.2.1  [23 nov 2010]

  *) The mifare_desfire_error_lookup() and mifare_desfire_get_last_error()
     functions were removed and replaced by the freefare_strerror(),
     freefare_strerror_r() and freefare_perror() functions.
  *) The library reports errors other that the ones returned by the PICC.
  *) The MDAD_KEYx macro where renamed MDAR_KEYx for consistency.
  *) The MDCM_MACING macro was renamed MDCM_MACED.
  *) The MDCM_FULLDES macro was renamed MDCM_ENCIPHERED.
  *) New function mifare_desfire_last_picc_error().
  *) New function mifare_desfire_aid_get_aid().

Changes between 0.1.0 and 0.2.0  [07 sep 2010]

  *) New function freefare_free_tag() to free individual tags from a MifareTag
     list.
  *) The mifare_application_alloc() size parameter is now expressed in bytes
     and not in sectors.
  *) New API function mad_sector_reserved().
  *) The mifare_classic_format_sector() functions now require a sector number
     instead of a block number.
  *) New API functions mifare_classic_block_sector(),
     mifare_classic_sector_first_block(), mifare_classic_sector_block_count()
     and mifare_classic_sector_last_block().
  *) New API functions mifare_application_read(), mifare_application_write().


libfreefare/freefare.c
#if defined(HAVE_CONFIG_H)
    #include "config.h"
#endif

#include <stdlib.h>
#include <string.h>

#include <freefare.h>

#include "freefare_internal.h"

#define MAX_CANDIDATES 16

#define NXP_MANUFACTURER_CODE 0x04

/*
 * Automagically allocate a FreefareTag given a device and target info.
 */
FreefareTag
freefare_tag_new(nfc_device *device, nfc_target target)
{
    FreefareTag tag = NULL;

    if (felica_taste(device, target)) {
	tag = felica_tag_new(device, target);
    } else if (mifare_mini_taste(device, target)) {
	tag = mifare_mini_tag_new(device, target);
    } else if (mifare_classic1k_taste(device, target)) {
	tag = mifare_classic1k_tag_new(device, target);
    } else if (mifare_classic4k_taste(device, target)) {
	tag = mifare_classic4k_tag_new(device, target);
    } else if (mifare_desfire_taste(device, target)) {
	tag = mifare_desfire_tag_new(device, target);
    } else if (ntag21x_taste(device, target)) {
	tag = ntag21x_tag_new(device, target);
    } else if (mifare_ultralightc_taste(device, target)) {
	tag = mifare_ultralightc_tag_new(device, target);
    } else if (mifare_ultralight_taste(device, target)) {
	tag = mifare_ultralight_tag_new(device, target);
    }

    return tag;
}


/*
 * MIFARE card common functions
 *
 * The following functions send NFC commands to the initiator to prepare
 * communication with a MIFARE card, and perform required cleannups after using
 * the targets.
 */

/*
 * Get a list of the MIFARE targets near to the provided NFC initiator.
 *
 * The list has to be freed using the freefare_free_tags() function.
 */
FreefareTag *
freefare_get_tags(nfc_device *device)
{
    FreefareTag *tags = NULL;
    int tag_count = 0;

    nfc_initiator_init(device);

    // Drop the field for a while
    nfc_device_set_property_bool(device, NP_ACTIVATE_FIELD, false);

    // Configure the CRC and Parity settings
    nfc_device_set_property_bool(device, NP_HANDLE_CRC, true);
    nfc_device_set_property_bool(device, NP_HANDLE_PARITY, true);
    nfc_device_set_property_bool(device, NP_AUTO_ISO14443_4, true);

    // Enable field so more power consuming cards can power themselves up
    nfc_device_set_property_bool(device, NP_ACTIVATE_FIELD, true);

    // Poll for a ISO14443A (MIFARE) tag
    nfc_target candidates[MAX_CANDIDATES];
    int candidates_count;
    nfc_modulation modulation = {
	.nmt = NMT_ISO14443A,
	.nbr = NBR_106
    };
    if ((candidates_count = nfc_initiator_list_passive_targets(device, modulation, candidates, MAX_CANDIDATES)) < 0)
	return NULL;

    tags = malloc(sizeof(void *));
    if (!tags) return NULL;
    tags[0] = NULL;

    for (int c = 0; c < candidates_count; c++) {
	FreefareTag t;
	if ((t = freefare_tag_new(device, candidates[c]))) {
	    /* (Re)Allocate memory for the found MIFARE targets array */
	    FreefareTag *p = realloc(tags, (tag_count + 2) * sizeof(FreefareTag));
	    if (p)
		tags = p;
	    else
		return tags; // FAIL! Return what has been found so far.
	    tags[tag_count++] = t;
	    tags[tag_count] = NULL;
	}
    }

    // Poll for a FELICA tag
    modulation.nmt = NMT_FELICA;
    modulation.nbr = NBR_424; // FIXME NBR_212 should also be supported
    if ((candidates_count = nfc_initiator_list_passive_targets(device, modulation, candidates, MAX_CANDIDATES)) < 0)
	return NULL;

    for (int c = 0; c < candidates_count; c++) {
	FreefareTag t;
	if ((t = freefare_tag_new(device, candidates[c]))) {
	    /* (Re)Allocate memory for the found FELICA targets array */
	    FreefareTag *p = realloc(tags, (tag_count + 2) * sizeof(FreefareTag));
	    if (p)
		tags = p;
	    else
		return tags; // FAIL! Return what has been found so far.
	    tags[tag_count++] = t;
	    tags[tag_count] = NULL;
	}
    }

    return tags;
}

/*
 * Returns the type of the provided tag.
 */
enum freefare_tag_type
freefare_get_tag_type(FreefareTag tag) {
    return tag->type;
}

/*
 * Returns the friendly name of the provided tag.
 */
const char *
freefare_get_tag_friendly_name(FreefareTag tag)
{
    switch (tag->type) {
    case FELICA:
	return "FeliCA";
    case MIFARE_MINI:
	return "Mifare Mini 0.3k";
    case MIFARE_CLASSIC_1K:
	return "Mifare Classic 1k";
    case MIFARE_CLASSIC_4K:
	return "Mifare Classic 4k";
    case MIFARE_DESFIRE:
	return "Mifare DESFire";
    case MIFARE_ULTRALIGHT_C:
	return "Mifare UltraLightC";
    case MIFARE_ULTRALIGHT:
	return "Mifare UltraLight";
    case NTAG_21x:
	return "NTAG21x";
    default:
	return "UNKNOWN";
    }
}

/*
 * Returns the UID of the provided tag.
 */
char *
freefare_get_tag_uid(FreefareTag tag)
{
    char *res = NULL;
    switch (tag->info.nm.nmt) {
    case NMT_FELICA:
	if ((res = malloc(17))) {
	    for (size_t i = 0; i < 8; i++)
		snprintf(res + 2 * i, 3, "%02x", tag->info.nti.nfi.abtId[i]);
	}
	break;
    case NMT_ISO14443A:
	if ((res = malloc(2 * tag->info.nti.nai.szUidLen + 1))) {
	    for (size_t i = 0; i < tag->info.nti.nai.szUidLen; i++)
		snprintf(res + 2 * i, 3, "%02x", tag->info.nti.nai.abtUid[i]);
	}
	break;
    case NMT_DEP:
    case NMT_ISO14443B2CT:
    case NMT_ISO14443B2SR:
    case NMT_ISO14443B:
    case NMT_ISO14443BI:
    case NMT_JEWEL:
    case NMT_BARCODE:
	res = strdup("UNKNOWN");
    }
    return res;
}

/*
 * Returns true if last selected tag is still present.
 */
bool freefare_selected_tag_is_present(nfc_device *device)
{
    return (nfc_initiator_target_is_present(device, NULL) == NFC_SUCCESS);
}

/*
 * Free the provided tag.
 */
void
freefare_free_tag(FreefareTag tag)
{
    if (tag) {
	tag->free_tag(tag);
    }
}

/*
 * Return lib version
 */
const char *
freefare_version(void)
{
    return PACKAGE_VERSION;
}


const char *
freefare_strerror(FreefareTag tag)
{
    const char *p = "Unknown error";
    if (nfc_device_get_last_error(tag->device) < 0) {
	p = nfc_strerror(tag->device);
    } else {
	if (tag->type == MIFARE_DESFIRE) {
	    if (MIFARE_DESFIRE(tag)->last_pcd_error) {
		p = mifare_desfire_error_lookup(MIFARE_DESFIRE(tag)->last_pcd_error);
	    } else if (MIFARE_DESFIRE(tag)->last_picc_error) {
		p = mifare_desfire_error_lookup(MIFARE_DESFIRE(tag)->last_picc_error);
	    }
	} else if (tag->type == NTAG_21x) {
	    p = ntag21x_error_lookup(NTAG_21x(tag)->last_error);
	}
    }
    return p;
}

int
freefare_strerror_r(FreefareTag tag, char *buffer, size_t len)
{
    return (snprintf(buffer, len, "%s", freefare_strerror(tag)) < 0) ? -1 : 0;
}

void
freefare_perror(FreefareTag tag, const char *string)
{
    fprintf(stderr, "%s: %s\n", string, freefare_strerror(tag));
}

/*
 * Free the provided tag list.
 */
void
freefare_free_tags(FreefareTag *tags)
{
    if (tags) {
	for (int i = 0; tags[i]; i++) {
	    freefare_free_tag(tags[i]);
	}
	free(tags);
    }
}


/*
 * Low-level API
 */

void *
memdup(const void *p, const size_t n)
{
    void *res;
    if ((res = malloc(n))) {
	memcpy(res, p, n);
    }
    return res;
}
